drivers/hmis_external_apis/app/views/hmis_external_apis/external_forms/form/_geolocation.haml

```diff
@@ -0,0 +1,101 @@
+:ruby
+  get_location_id = next_html_id
+  clear_id = next_html_id
+  text_id = next_html_id
+  map_id = next_html_id
+  input_id = next_html_id
+%label{for:input_id, class: required && 'required'}= label
+%br
+%button.btn.btn-secondary.mb-2{id: get_location_id}
+  Get location
+%button.btn.btn-secondary.mb-2{id: clear_id}
+  Clear
+.form-text.text-muted.mb-2{id: text_id}
+  Loading...
+%div.leaflet-container{id: map_id}
+%input{id: input_id, hidden: true, name: name}
+
+- content_for :page_js do
+  :javascript
+    // 'use strict';
+    $(function() {
+      var helperTextId = '#' + #{text_id.to_json.html_safe};
+      var getLocationButtonId = '#' + #{get_location_id.to_json.html_safe};
+      var clearButtonId = '#' + #{clear_id.to_json.html_safe};
+      var mapContainerId = '#' + #{map_id.to_json.html_safe};
+      var inputId = '#' + #{input_id.to_json.html_safe}
+
+      var helperText = $(helperTextId);
+      var getLocationButton = $(getLocationButtonId);
+      var clearButton = $(clearButtonId);
+      var mapContainer = $(mapContainerId);
+      var input = $(inputId);
+
+      helperText.hide();
+      mapContainer.hide();
+      clearButton.hide();
+
+      var map = L.map(#{map_id.to_json.html_safe});
+      L.tileLayer('https://tile.openstreetmap.org/{z}/{x}/{y}.png', {
+        maxZoom: 19,
+        attribution: '&copy; <a href="http://www.openstreetmap.org/copyright">OpenStreetMap</a>'
+      }).addTo(map);
+      var marker;
+
+      if (!navigator.geolocation) {
+        helperText.show();
+        helperText.text("Geolocation is not supported by this browser.");
+        getLocationButton.prop('disabled', true)
+        input.val('unavailable');
+      }
+
+      var setPosition = function(position) {
+        setLoading(false);
+        mapContainer.show();
+        clearButton.show();
+
+        var lat = position.coords.latitude;
+        var lon = position.coords.longitude;
+
+        // 13 is zoom level, consistent with WH MapWithMarkers
+        map.setView([lat, lon], 13);
+        if (marker) {
+          marker.remove();
+        }
+        marker = L.marker([lat, lon]).addTo(map);
+        input.val(JSON.stringify({ "latitude": lat, "longitude": lon }));
+      }
+
+      var onError = function(error) {
+        getLocationButton.prop('disabled', false);
+        mapContainer.hide();
+        clearButton.hide();
+        helperText.text("Location unavailable.", error.message);
+        input.val('unavailable');
```

Comment:
This is an incomplete first pass at the desired "required" behavior that we discussed yesterday:
- (if geolocation is available in this browser) clicking the button is required
- but the user is allowed to clear and still submit

Reply:
Thanks Martha. Sorry we didn;t get a chance to sync up on that- my takeaway was that it would also be _fine_ to have the geolocation be optional, because they could rely on training to make sure folks use it. I think that is probably the safer option–– safer as in, we less chance of someone being unable to submit forms because of device issues. Thoughts?

Reply:
Sounds good. I think either behavior is supported right now. If the item has `required` set to true, then the behavior is what I described in my earlier comment. Otherwise geolocation is fully optional.

Reply:
great 👍